pkg/apis/extensions/latest/latest.go

```diff
@@ -0,0 +1,105 @@
+/*
+Copyright 2014 The Kubernetes Authors All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package latest
```

Comment:
I agree with your assertion about code that has a "true, hard dependency".
However, this code is not (or at least should not be) well described by
that. We depend on a codec existing, but we should not be opinionated about
what package provides that codec. (If it's only possible to construct a
working codec with the default install package, then that is a different
code problem!)

On Tue, Dec 8, 2015 at 4:43 AM, David Eads <notifications@github.com> wrote:

> In pkg/apis/extensions/latest/latest.go
> <https://github.com/kubernetes/kubernetes/pull/18154#discussion_r46948023>
> :
>
> > +Copyright 2014 The Kubernetes Authors All rights reserved.
> > +
> > +Licensed under the Apache License, Version 2.0 (the "License");
> > +you may not use this file except in compliance with the License.
> > +You may obtain a copy of the License at
> > +
> > +    http://www.apache.org/licenses/LICENSE-2.0
> > +
> > +Unless required by applicable law or agreed to in writing, software
> > +distributed under the License is distributed on an "AS IS" BASIS,
> > +WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
> > +See the License for the specific language governing permissions and
> > +limitations under the License.
> > +*/
> > +
> > +package latest
>
> You say "two" above. We will have dozens of groups or more, and not all of
> them will live in the main repo. Do you still think this is a good pattern?
>
> Yes. I think that code that is designed to work with exactly one specific
> group, that will fail hard without that group, and is only activated when
> that group is present should explicitly depend upon its dependencies to let
> the compiler help enforce its work.
>
> Of the 99 uses in the code, 97 of them fall into the case I just described
> (one). Two of those uses are "correct" (infinity).
>
> I don't think code that has a true, hard dependency is improved by hiding
> that dependency behind a map lookup.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/18154/files#r46948023>.
>